library/core/src/str/validations.rs

```diff
@@ -112,162 +112,290 @@ where
-                    // always aligned with a `usize` so it's safe to dereference
-                    // both `block` and `block.add(1)`.
-                    unsafe {
-                        let block = ptr.add(index) as *const usize;
-                        // break if there is a nonascii byte
-                        let zu = contains_nonascii(*block);
-                        let zv = contains_nonascii(*block.add(1));
-                        if zu || zv {
-                            break;
-                        }
-                    }
-                    index += ascii_block_size;
-                }
-                // step from the point where the wordwise loop stopped
-                while index < len && v[index] < 128 {
-                    index += 1;
-                }
-            } else {
-                index += 1;
+            // non-ASCII case: validate up to 4 bytes, then advance `curr`
+            // accordingly
+            match validate_non_acii_bytes(buf, curr) {
+                Ok(next) => curr = next,
+                Err(e) => return Err(e),
             }
         }
     }
 
     Ok(())
 }
 
-// https://tools.ietf.org/html/rfc3629
-const UTF8_CHAR_WIDTH: &[u8; 256] = &[
-    // 1  2  3  4  5  6  7  8  9  A  B  C  D  E  F
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 0
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 1
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 2
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 3
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 4
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 5
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 6
-    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 7
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 8
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 9
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // A
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // B
-    0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // C
-    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, // D
-    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, // E
-    4, 4, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // F
-];
+#[inline]
+const fn validate_non_acii_bytes(buf: &[u8], mut curr: usize) -> Result<usize, Utf8Error> {
+    const fn subarray<const N: usize>(buf: &[u8], idx: usize) -> Option<[u8; N]> {
+        if buf.len() - idx < N {
+            return None;
+        }
+
+        // SAFETY: checked in previous condition
+        Some(unsafe { *(buf.as_ptr().add(idx) as *const [u8; N]) })
+    }
+
+    let prev = curr;
+    macro_rules! err {
+        ($error_len: expr) => {
+            return Err(Utf8Error { valid_up_to: prev, error_len: $error_len })
+        };
+    }
+
+    let b0 = buf[curr];
+    match utf8_char_width(b0) {
+        2 => {
+            let Some([_, b1]) = subarray(buf, curr) else {
+                err!(None);
+            };
```

Comment:
> Have you tried benching a simpler version that doesn't use the 4-word and 8-word blocks, or indeed just skips "words" altogether, and instead uses 16-byte alignment unconditionally? We have several 32-bit targets that still have the capability to use 128-bit hardware vectors and still prefer vector alignment when doing so. This kind of ASCII-specific fast-path should be able to do more than a 2x speedup, though it might be hard to accomplish when working mostly with autovectorization.

Yes, I have benchmarked various combinations of different block sizes, and this variant came up on top in most of the test cases. For 32-bit targets, this would correspond to the 4-word block size, so this should be covered, although I have not tried using explicit (target-independent) block sizes. You would have to use byte-wise operations exclusively, probably obfuscating the code a bit but the compiler should be still be able to auto-vectorize.